File: packages/python/dynamic/runners/langchain/agent.py
import asyncio
from dataclasses import dataclass
from typing import Union, Dict, Optional

from fastapi import WebSocket

# dyanmic
from dynamic.runners.runner import Runner
from dynamic.classes.dynamic_agent  import DynamicAgent
from dynamic.runners.langchain.config import ChainRunnerConfig

# langchain
from langchain.agents import Agent, AgentExecutor, initialize_agent


class AgentRunner(Runner):
    def __init__(self,
        handle: Union[DynamicAgent, Agent, AgentExecutor],
        config: ChainRunnerConfig,
        websocket: Optional[WebSocket] = None,
        streaming: bool = False,
    ):
        self.streaming = streaming
        self.config = config

        if streaming:
            if not isinstance(handle, DynamicAgent):
                raise ValueError(f"A streaming Agent needs to a DynamicAgent, recieved {type(handle)}.")

            handle = handle._initialize_agent_with_websocket(websocket)

        if not (isinstance(handle, Agent) or isinstance(handle, AgentExecutor)):
            raise ValueError(f"AgentRunner requires handle to be a Langchain Agent or AgentExecutor. Instead got {type(handle)}.")
        
        super(AgentRunner, self).__init__(handle, config)

    async def arun(self):
        input = self.config.input
        if self.streaming:
            return await self.handle.arun(input)

        return self.handle.run(input)
    
    def run(self):
        input = self.config.input
        return self.handle.run(input)

File: packages/js/dynamic/src/runners/langchain/agent.ts
// dynamic
import { Runner, type Config } from '../types';
import { DynamicAgent } from '../../dync/langchain/agent';
import { type AgentExecutor } from 'langchain/agents';

export class AgentRunner extends Runner {
  websocket?: WebSocket;
  streaming: boolean; // marker to indicate agent will stream its tokens

  constructor(
    handle: DynamicAgent | AgentExecutor,
    config: Config,
    websocket?: WebSocket,
    streaming?: boolean
  ) {
    super(handle, config);
    this.streaming = streaming ?? false;
    if (this.streaming) {
      if (!(this.handle instanceof DynamicAgent)) {
        throw new Error(
          `"handle" expected DynamicAgent, recieved ${typeof this.handle}`
        );
      }
      if (websocket == null) {
        throw new Error(
          'DynamicAgent marked as `streaming` without a websocket.'
        );
      }
      this.streaming = true;
      this.websocket = websocket;
    }
  }

  run(): any {
    const { input } = this.config;
    return this.handle.run(input);
  }

  async arun(): Promise<any> {
    const { streaming, websocket } = this;
    if (!(websocket && streaming) || !(this.handle instanceof DynamicAgent)) {
      throw new Error(
        'This is not a streaming agent, please use run() and not arun().'
      );
    }
    const handle = await this.handle.initAgentWithWebSocket(websocket);

    const { input } = this.config;
    return await handle.run(input);
  }
}
